modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -535,4 +612,147 @@ private static void validateParsedStatement(
             }
         }
     }
+
+    private class MultiStatementHandler {
+        private final String schemaName;
+        private final IgniteTransactions transactions;
+        private final @Nullable InternalTransaction explicitTransaction;
+        private final Queue<ScriptStatementParameters> statements;
+
+        MultiStatementHandler(
+                String schemaName,
+                IgniteTransactions transactions,
+                @Nullable InternalTransaction explicitTransaction,
+                List<ParsedResult> parsedResults,
+                Object[] params
+        ) {
+            this.schemaName = schemaName;
+            this.transactions = transactions;
+            this.explicitTransaction = explicitTransaction;
+            this.statements = prepareStatementsQueue(parsedResults, params);
+        }
+
+        CompletableFuture<AsyncSqlCursor<List<Object>>> processNext() {
+            ScriptStatementParameters parameters = statements.poll();
+
+            if (parameters == null) {
+                return CompletableFuture.completedFuture(null);
+            }
+
+            ParsedResult parsedResult = parameters.parsedResult;
+            Object[] dynamicParams = parameters.dynamicParams;
+            CompletableFuture<AsyncSqlCursor<List<Object>>> cursorFuture = parameters.cursorFuture;
+
+            try {
+                if (cursorFuture.isDone()) {
+                    return cursorFuture;
+                }
+
+                QueryTransactionWrapper txWrapper = wrapTxOrStartImplicit(parsedResult.queryType(), transactions, explicitTransaction);
```

Comment:
Can we replace both agruments with a single QueryTransactionWrapper
```
IgniteTransactions transactions,
@Nullable InternalTransaction explicitTransaction,
```
Then we could make use a pattern
```
txWrapper.startTxIfNeeded(queryType);
... execute query ... 
txWrapper.commitImplicit();
```
Inside the wrapper, we can have all the tx logic in one place and ensure 
* a query type match a tx type
* no inner transactions started
* prev. tx was commited/rolled back
* ignore tx commands in case of explicit tx.

Reply:
WDYT?

Reply:
Overall, I agree with you.

But since I remember that there were a LOT of discussions on this topic previously [1] and since this task is not directly related to current issue? I filled out another ticket to resolve this separately [2]

[1] https://github.com/apache/ignite-3/pull/2449
[2] https://issues.apache.org/jira/browse/IGNITE-20855